lib/SILGen/SILGenApply.cpp

```diff
@@ -6067,6 +6067,39 @@ RValue SILGenFunction::emitDynamicSubscriptExpr(DynamicSubscriptExpr *e,
   return RValue(*this, e, emitManagedRValueWithCleanup(optResult, optTL));
 }
 
+SmallVector<ManagedValue, 4> SILGenFunction::emitKeyPathSubscriptOperands(
+    SubscriptDecl *subscript, SubstitutionMap subs, Expr *indexExpr,
+    CanSILFunctionType fnType) {
```

Comment:
I believe the problem you're running into with generic parameters is due to the fact that you're passing the function type for the key path accessor here, which will contain generic parameter types. However the the original parameter types you're supplying (`origFnType`) have been substituted with the replacement types.

As far as I can tell, we just want to be emitting the arguments directly without performing any abstraction changes or things like that, so I *think* what we want to do is remove the `fnType` parameter here, and lower the concrete type of the function ourselves, e.g `auto fnType = getLoweredType(origFnType, substFnType).castTo<SILFunctionType>()`. That appears to work at least for the generic example you gave.

Reply:
Fantastic! This seems to pass all the tests.

Reply:
Great! Could you add some FileCheck SILGen tests alongside your existing tests in test/SILGen/keypaths.swift to make sure we emit the correct SILGen for a few cases? I think we'll want a test for at least:

- Your example with a generic default
- An example with a caller-side default such as `#function`
- An example with a generic caller-side default, e.g `subscript<T : ExpressibleByStringLiteral>(x x: T = #function, y y: T) -> T { x }`, called using `\Foo.[y: ""]`

You can take a look at:

https://github.com/apple/swift/blob/cd3ada5abbe78d4062b7d323e5fd15a2df3d3524/test/SILGen/default_arguments.swift#L44-L57

to see an example of a FileCheck test for the SILGen for a regular default argument.

It would be great if you could also add a test case for a cross-module default.

Reply:
I tried to write a FileCheck test but it can't match the string. I suspect its a simple issue. Any ideas?
**Input**
```swift
// CHECK-LABEL: sil hidden [ossa] @{{.*}}check_default_subscripts
func check_default_subscripts() {
  // CHECK: [[INTINIT:%[0-9]+]] = integer_literal $Builtin.Int64, 0
  // CHECK: [[I:%[0-9]+]] = struct $Int ([[INTINIT]] : $Builtin.Int64)
  // CHECK: [[DFN:%[0-9]+]] = function_ref @$s3run18SubscriptDefaults4V1x1yxx_xtcSjRzluipfA0_ : $@convention(thin) <τ_0_0 where τ_0_0 : Numeric> () -> @out τ_0_0
  // CHECK: [[ALLOC:%[0-9]+]] = alloc_stack $Int
  // CHECK: apply [[DFN]]<Int>([[ALLOC]]) : $@convention(thin) <τ_0_0 where τ_0_0 : Numeric> () -> @out τ_0_0
  // CHECK: [[LOAD:%[0-9]+]] = load [[ALLOC]] : $*Int
  // CHECK: [[KEYPATH:%[0-9]+]] = keypath $WritableKeyPath<SubscriptDefaults4, Int>, (root $SubscriptDefaults4; settable_property $Int,  id @$s3run18SubscriptDefaults4V1x1yxx_xtcSjRzluig : $@convention(method) <τ_0_0 where τ_0_0 : Numeric> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, SubscriptDefaults4) -> @out τ_0_0, getter @$s3run18SubscriptDefaults4V1x1yxx_xtcSjRzluipACSiTK : $@convention(thin) (@in_guaranteed SubscriptDefaults4, UnsafeRawPointer) -> @out Int, setter @$s3run18SubscriptDefaults4V1x1yxx_xtcSjRzluipACSiTk : $@convention(thin) (@in_guaranteed Int, @inout SubscriptDefaults4, UnsafeRawPointer) -> (), indices [%$0 : $Int : $Int, %$1 : $Int : $Int], indices_equals @$sS2iTH : $@convention(thin) (UnsafeRawPointer, UnsafeRawPointer) -> Bool, indices_hash @$sS2iTh : $@convention(thin) (UnsafeRawPointer) -> Int) ([[I]], [[LOAD]])
  _ = \SubscriptDefaults4.[x: 0]
}
```
**Error**
```
swift/test/SILGen/keypaths.swift:423:12: error: CHECK: expected string not found in input
 // CHECK: [[INTINIT:%[0-9]+]] = integer_literal $Builtin.Int64, 0
           ^
<stdin>:5700:57: note: scanning from here
sil hidden [ossa] @$s8keypaths24check_default_subscriptsyyF : $@convention(thin) () -> () {
                                                        ^
<stdin>:5702:2: note: possible intended match here
 %0 = integer_literal $Builtin.IntLiteral, 0 // user: %3
```
I've looked and the emitted silgen _does_ contian `integer_literal $Builtin.Int64, 0`:
```
// check_default_subscripts()
sil hidden @$s3run24check_default_subscriptsyyF : $@convention(thin) () -> () {
bb0:
  %0 = integer_literal $Builtin.Int64, 0          // user: %1
  %1 = struct $Int (%0 : $Builtin.Int64)          // user: %6
...
```

Thanks again for the help!

Reply:
@zoecarver Ah, I think you're looking at output from `-emit-sil` instead of `-emit-silgen`, the test uses the latter so that's what you'll want to use for reference.

Reply:
@hamishknight yep, that's it!

Reply:
@hamishknight sorry for the delay. I think I've got all the test cases you asked for. Let me know if I missed any, though.